### ISSUE TITLE
[RFC]: Custom Scheduler Injection for Diffusion Pipelines

### ISSUE THREAD

**Issue by knlnguyen1802** [vllm-project/vllm-omni#6269]
### Motivation.

Every diffusion pipeline in vllm-omni **hardcodes its scheduler construction**. Today there are
five such sites — `flux`, `qwen_image`, `sd3`, `ltx2` call
`FlowMatchEulerDiscreteScheduler.from_pretrained(model, subfolder="scheduler")` directly in
`__init__` (or the component loader), and `wan2_2` uses a private builder
(`build_wan_scheduler`). There is no supported way to run a pipeline with a scheduler the
pipeline author did not anticipate.

This matters most for **RL post-training of diffusion models** (FlowGRPO, DanceGRPO, MixGRPO,
DPO/NFT-style methods). These algorithms need the denoise loop to run an **SDE variant** of the
flow-matching scheduler — one that injects noise at chosen steps and computes per-step
log-probabilities of the sampled latents. This is an *algorithm-level* concern: the SDE
formulas, noise levels, and sampling windows belong to the RL method, not to the serving
engine. The denoise loop itself — transformer calls, CFG, batching, VAE decode — is completely
unchanged.

Because the scheduler cannot be swapped, downstream RL frameworks today have exactly one
escape hatch: **rewrite the whole pipeline**. In verl-omni this means 7 custom pipeline
adapters (qwen_image, sd3, ltx2, wan2_2, edit, mix, bagel), each copying the upstream denoise
loop — up to ~900 lines per adapter — just to change the ~5 lines that call `scheduler.step`.
The result is structurally fragile:

- Each adapter must track upstream loop refactors bit-for-bit; any drift silently breaks
  numerical parity between rollout and training.
- verl-omni's vllm-omni pin has fallen hundreds of commits behind, and each upgrade is a
  multi-file reconciliation effort (the last one touched ~30 files).
- Upstream field renames/removals in the output path break every adapter at once.

**The only thing RL actually needs to change is the scheduler.** Everything else should stay
upstream.

The beneficiaries are not limited to RL. A scheduler seam also serves:

- **Alternative samplers** — DPM/UniPC variants, distilled/few-step schedulers, research
  samplers — without forking pipelines.
- **Determinism-sensitive serving** — schedulers with explicit RNG control for reproducible
  generation.
- **A/B evaluation** — swapping solvers via config instead of code edits.

## Design

Three additive pieces. **No denoise-loop edits anywhere.**

### 1. Scheduler registry + factory

New module `vllm_omni/diffusion/models/schedulers/registry.py` (co-located with the existing
scheduler implementations; re-exported from the package `__init__`):

```python
def register_scheduler(name: str, cls: type | None = None):
    """Direct call or decorator: @register_scheduler("flow_match_sde")."""

def resolve_scheduler_cls(ref: str | type | None) -> type | None:
    """Accepts a registry name, a dotted import path, a class, or None."""

def build_pipeline_scheduler(od_config, scheduler_cls=None,
                             scheduler_kwargs=None, default_builder=None):
    """Resolution order: explicit arg > od_config.scheduler > pipeline default."""
```

Registry entries load lazily from the `vllm_omni.schedulers` entry-point group, so installing
an external package (e.g. verl-omni) makes its schedulers available by name — no vllm-omni
code changes.

### 2. Configuration

Two new fields on `OmniDiffusionConfig` (threaded through engine args, mirroring
`custom_pipeline_args`):

```python
scheduler: str | None = None                 # registry name or dotted path
scheduler_kwargs: dict[str, Any] | None = None
```

```python
# Python API
AsyncOmni(model="Qwen/Qwen-Image", scheduler="flow_match_sde")
```

### 3. Call-site swaps (5, mechanical)

Each hardcoded construction site routes through the factory, passing its current construction
as `default_builder` — so the un-configured path executes the identical code as today:

```python
# before (pipeline_qwen_image.py)
self.scheduler = FlowMatchEulerDiscreteScheduler.from_pretrained(
    model, subfolder="scheduler", local_files_only=local_files_only)

# after
self.scheduler = build_pipeline_scheduler(
    od_config,
    default_builder=lambda: FlowMatchEulerDiscreteScheduler.from_pretrained(
        model, subfolder="scheduler", local_files_only=local_files_only))
```

Sites: `flux`, `qwen_image`, `sd3`, `ltx2` (component loader), `wan2_2` (init + its existing
per-request solver/shift rebuild, which is skipped with a warning when an injected scheduler
is active — per-request knob handling belongs to the follow-up protocol).

### Scheduler contract

Injected classes must satisfy a small, documented, diffusers-compatible contract:

- `step(noise_pred, t, latents, return_dict=False, generator=None)`
- `set_begin_index(...)` and `set_timesteps(...)`
- a `.config` attribute
- deepcopy-safe (step-wise execution deep-copies the scheduler per request)

All target loops already funnel through `scheduler_step_maybe_with_cfg`, which accepts
`per_request_scheduler` — so the injected object flows through existing plumbing with zero
loop changes.

### Feedback Period.

1 week

### CC List.

@SamitHuang 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.
